lib/testutils/src/lib.rs

```diff
@@ -383,61 +383,136 @@ pub fn write_file(store: &Store, path: &RepoPath, contents: &str) -> FileId {
+        TestTreeFileEntryBuilder {
+            tree_builder: self.tree_builder,
+            path: self.path,
+            contents,
             executable: false,
-        },
-    );
-    id
+        }
+    }
+
+    pub fn text_file(self, contents: &str) -> TestTreeFileEntryBuilder<'a> {
+        self.binary_file(contents.as_bytes().to_vec())
+    }
+
+    pub fn empty_file(self) -> TestTreeFileEntryBuilder<'a> {
+        self.binary_file(vec![])
+    }
+
+    pub fn symlink(self, target: &str) {
+        let TestTreeEntryBuilder { tree_builder, path } = self;
+        let id = tree_builder
+            .store()
+            .write_symlink(&path, target)
+            .block_on()
+            .unwrap();
+        tree_builder.set(path, TreeValue::Symlink(id));
+    }
+
+    pub fn submodule(self, commit: CommitId) {
+        let TestTreeEntryBuilder { tree_builder, path } = self;
+        tree_builder.set(path, TreeValue::GitSubmodule(commit));
+    }
 }
 
-pub fn write_executable_file(tree_builder: &mut TreeBuilder, path: &RepoPath, contents: &str) {
-    let id = write_file(tree_builder.store(), path, contents);
-    tree_builder.set(
-        path.to_owned(),
-        TreeValue::File {
-            id,
-            executable: true,
-        },
-    );
+pub struct TestTreeFileEntryBuilder<'a> {
+    tree_builder: &'a mut TreeBuilder,
+    path: RepoPathBuf,
+    contents: Vec<u8>,
+    executable: bool,
 }
 
-pub fn write_symlink(tree_builder: &mut TreeBuilder, path: &RepoPath, target: &str) {
-    let id = tree_builder
-        .store()
-        .write_symlink(path, target)
-        .block_on()
-        .unwrap();
-    tree_builder.set(path.to_owned(), TreeValue::Symlink(id));
+impl TestTreeFileEntryBuilder<'_> {
+    pub fn executable(mut self, executable: bool) -> Self {
+        self.executable = executable;
+        self
+    }
 }
 
-pub fn create_single_tree(repo: &Arc<ReadonlyRepo>, path_contents: &[(&RepoPath, &str)]) -> Tree {
-    let store = repo.store();
-    let mut tree_builder = store.tree_builder(store.empty_tree_id().clone());
-    for (path, contents) in path_contents {
-        write_normal_file(&mut tree_builder, path, contents);
+impl Drop for TestTreeFileEntryBuilder<'_> {
+    fn drop(&mut self) {
+        let id = self
+            .tree_builder
+            .store()
+            .write_file(&self.path, &mut self.contents.as_slice())
+            .block_on()
+            .unwrap();
+        let path = std::mem::replace(&mut self.path, RepoPathBuf::root());
+        self.tree_builder.set(
+            path,
+            TreeValue::File {
+                id,
+                executable: self.executable,
+            },
+        );
     }
-    let id = tree_builder.write_tree().unwrap();
-    store.get_tree(RepoPathBuf::root(), &id).unwrap()
 }
 
-pub fn create_tree(repo: &Arc<ReadonlyRepo>, path_contents: &[(&RepoPath, &str)]) -> MergedTree {
```

Comment:
nit: just leave the current `create_tree()` for convenience, and add `create_tree_with()` or something for advanced use case?